src/edu/stanford/nlp/sempre/Grammar.java

```diff
@@ -271,24 +313,58 @@ private void interpretRule(LispTree tree) {
     addRule(rule, isOptionals);
   }
 
-  public synchronized boolean addRule(Rule rule) {
+  public synchronized void addRule(Rule rule) {
     List<Boolean> isOptionals = new ArrayList<>();
-    for (String rhs : rule.rhs) isOptionals.add(false);
-    return addRule(rule, isOptionals);
+    for (int i = 0; i < rule.rhs.size(); i++) isOptionals.add(false);
+    addRule(rule, isOptionals);
   }
 
   // Add a rule to the grammar.
-  public synchronized boolean addRule(Rule rule, List<Boolean> isOptionals) {
+  public synchronized void addRule(Rule rule, List<Boolean> isOptionals) {
     rules.addAll(binarizeRule(rule, isOptionals));
-    return true;
+  }
+
+  private void collectValidTags(LispTree tree) {
+    if (tree.isLeaf())
+      throw new RuntimeException("Expected list, got " + tree);
+    try {
+      String command = tree.child(0).value;
+      if ("when".equals(command)) {
+        collectValidTagsBoolean(tree.child(1));
+        for (int i = 2; i < tree.children.size(); i++)
+          collectValidTags(tree.child(i));
+      } else if ("for".equals(command)) {
+        for (int i = 3; i < tree.children.size(); i++)
+          collectValidTags(tree.child(i));
+      }
+    } catch (Exception e) {
+      e.printStackTrace();
+      throw new RuntimeException("Error on " + tree + ": " + e);
+    }
+  }
+
+  private void collectValidTagsBoolean(LispTree tree) {
+    if (tree.isLeaf()) {
+      validTags.add(tree.value);
+      return;
+    }
+    String command = tree.child(0).value;
+    if ("not".equals(command))
+      collectValidTagsBoolean(tree.child(1));
+    else if ("and".equals(command) || "or".equals(command))
+      for (int i = 1; i < tree.children.size(); i++)
+        collectValidTagsBoolean(tree.child(i));
   }
 
   // Generate intermediate categories for binarization.
+  public static final String INTERMEDIATE_PREFIX = "$Intermediate";
   private int freshCatIndex = 0;
   private String generateFreshCat() {
     freshCatIndex++;
-    return "$Intermediate" + freshCatIndex;
-    // return "$I" + freshCatIndex + ":" + content;
```

Comment:
Ideally we would use "$I" + freshCatIndex + ":" + content, since these categories are more informative.  We didn't change it before to preserve backward compatibility.  But maybe this is a good time to switch things over to use "$I"...?